vignettes/python-primer.Rmd

```diff
@@ -0,0 +1,920 @@
+execution, it preserves the state of everything in the function body and
+returns control to whatever is iterating over the generator object.
+Calling `next()` on the generator object resumes execution of the
+function body until the next `yield` is encountered, or the function
+finishes.
+
+### Iteration closing remarks
+
+Iteration is deeply baked into the Python language, and R users may be
+surprised by how things in Python are iterable, iterators, or powered by
+the iterator protocol under the hood. For example, the built-in `map()`
+$$equivalent to R's `lapply()`$$ yields an iterator, not a list.
+Similarly, a tuple comprehension like `(elem for elem in x)` produces an
+iterator. Most features dealing with files are iterators, and so on. Any
+time you find an iterator inconvenient, you can materialize all the
+elements into a list using the Python built-in `list()`, or
+`reticulate::iterate()` in R.
+
+### Integers and Floats
+
+R users generally don't need to be aware of the difference between
+integers and floating point numbers, but that's not the case in Python.
+If this is your first exposure to numeric data types, here are the
+essentials:
+
+-   integer types can only represent whole numbers like `1` or `2`, not
+    floating point numbers like `1.2`.
+
+-   floating-point types can represent any number, but with some degree
+    of imprecision.
+
+In R, writing a bare literal number like `12` produces a floating point
+type, whereas in Python, it produces an integer. You can produce an
+integer literal in R by appending an `L`, as in `12L`. Many Python
+functions expect integers, and will error when provided a float.
+
+For example, say we have a Python function that expects an integer:
+
+```{python}
+def a_strict_Python_function(x):
+  assert isinstance(x, int), "x is not an int"
+  print("Yay! x was an int")
+```
+
+When calling it from R, you must be sure to call it with an integer:
+
+```{r}
+library(reticulate)
+py$a_strict_Python_function(3) |> try()    # error
+py$a_strict_Python_function(3L)            # success
+py$a_strict_Python_function(as.integer(3)) # success
+```
+
+### What about R vectors?
+
+R is a language designed for numerical computing first. Vector data
+types are baked deep into the R language, to the point that the language
+doesn't even distinguish scalars from vectors. By comparison, numerical
+computing features in Python are generally provided by third party
+packages (*modules*, in Python parlance). The closest equivalent to an R
+numeric vector is a numpy array, or sometimes, a Python list of numerics
+(some Pythonistas might argue for `array.array()` here, but that's so
+rarely encountered in actual Python code we don't mention it further).
+Teaching the NumPy interface is beyond the scope of this tutorial, but
+it's worth pointing out some potential tripping hazards for users
+accustomed to R arrays:
+
+-   When indexing into multidimensional numpy arrays, trailing
```

Comment:
I was wondering if it'd make sense to even start talking about numpy here ... on the other hand, that point is totally worth making, - again one super confusing thing when seen from "outside" ...

Reply:
Thanks! Kevin also honed in on the way Numpy is introduced. I re-ordered sections so that module are talked about first, before vectors, and I word-smithed this a little too.